File: go/sb/rpc.go
package sb

import (
	"bytes"
	"context"
	"io"
	"net/http"
	"time"
)

type RpcErrCode int

const (
	RpcOk        RpcErrCode = 200
	RpcNoConn    RpcErrCode = 0
	RpcTimeout   RpcErrCode = 408
	RpcReqErr    RpcErrCode = 400
	RpcRespErr   RpcErrCode = 500
	RpcNotAuth   RpcErrCode = 401
	RpcNotExist  RpcErrCode = 404
)

type Client struct {
	BaseURL string
	HTTP    *http.Client
	Timeout time.Duration
	Retries int
	headers map[string]string
}

func NewClient(baseURL string) *Client {
	return &Client{
		BaseURL: baseURL,
		HTTP:    &http.Client{},
		Timeout: 5 * time.Second,
		Retries: 3,
		headers: make(map[string]string),
	}
}

func SetClientHeader(c *Client, key, value string) {
	if c == nil { return }
	if c.headers == nil { c.headers = make(map[string]string) }
	c.headers[key] = value
}
func GetClientHeader(c *Client, key string) string {
	if c == nil { return "" }
	return c.headers[key]
}
func RemoveClientHeader(c *Client, key string) {
	if c == nil { return }
	delete(c.headers, key)
}

func SetClientAuthorization(c *Client, token string) { SetClientHeader(c, "Authorization", "Bearer "+token) }
func GetClientAuthorization(c *Client) string        { return GetClientHeader(c, "Authorization") }
func RemoveClientAuthorization(c *Client)            { RemoveClientHeader(c, "Authorization") }
func IsClientAuthorized(c *Client) bool              { return GetClientAuthorization(c) != "" }

func isTimeout(err error) bool {
	if err == nil { return false }
	if err == context.DeadlineExceeded { return true }
	if netErr, ok := err.(interface{ Timeout() bool }); ok && netErr.Timeout() { return true }
	return false
}

func doClient(c *Client, ctx context.Context, path string, body []byte) ([]byte, RpcErrCode) {
	if c == nil { return nil, RpcNoConn }
	if c.HTTP == nil { c.HTTP = &http.Client{} }
	var resp *http.Response
	var err error

	for i := 0; i <= c.Retries; i++ {
		if i > 0 {
			timer := time.NewTimer(time.Duration(i) * time.Second)
		select {
		case <-ctx.Done():
			timer.Stop()
			return nil, RpcTimeout
		case <-timer.C:
		}
	}

		reqCtx := ctx
		cancel := func() {}
		if c.Timeout > 0 {
			reqCtx, cancel = context.WithTimeout(ctx, c.Timeout)
		}
		req, reqErr := http.NewRequestWithContext(reqCtx, "POST", c.BaseURL+path, bytes.NewReader(body))
		if reqErr != nil {
			cancel()
			return nil, RpcNoConn
		}

		for k, v := range c.headers {
			req.Header.Set(k, v)
		}

		resp, err = c.HTTP.Do(req)
		cancel()
		if err != nil {
			if isTimeout(err) && i < c.Retries {
				continue
			}
			if isTimeout(err) {
				return nil, RpcTimeout
			}
			return nil, RpcNoConn
		}

		if resp.StatusCode == http.StatusRequestTimeout && i < c.Retries {
			resp.Body.Close()
			continue
		}
		break
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, RpcErrCode(resp.StatusCode)
	}

	b, readErr := io.ReadAll(resp.Body)
	if readErr != nil {
		return nil, RpcRespErr
	}
	return b, RpcOk
}

// CallUserGetAbc 获取用户的id
func CallUserGetAbc(c *Client, ctx context.Context) (result OrderStatus, errCode RpcErrCode) {
	var buf bytes.Buffer

	body, status := doClient(c, ctx, "/user.get_abc", buf.Bytes())
	if status != RpcOk {
		return result, status
	}

	if err := GetAll(bytes.NewBuffer(body), func(buf *bytes.Buffer) error {
		val, err := GetOrderStatus(buf)
		if err != nil { return err }
		result = val
		return nil
	}); err != nil {
		return result, RpcRespErr
	}
	return result, status
}
// CallUserGetAbcd 获取abcd
func CallUserGetAbcd(c *Client, ctx context.Context, page uint8, size uint8) (result OrderStatus, errCode RpcErrCode) {
	var buf bytes.Buffer
	if err := SetAll(&buf, func(buf *bytes.Buffer) error {
		return SetU8(buf, page)
	}, func(buf *bytes.Buffer) error {
		return SetU8(buf, size)
	}); err != nil {
		return result, RpcReqErr
	}

	body, status := doClient(c, ctx, "/user.get_abcd", buf.Bytes())
	if status != RpcOk {
		return result, status
	}

	if err := GetAll(bytes.NewBuffer(body), func(buf *bytes.Buffer) error {
		val, err := GetOrderStatus(buf)
		if err != nil { return err }
		result = val
		return nil
	}); err != nil {
		return result, RpcRespErr
	}
	return result, status
}
// CallUserSetSimInfo 设置sim信息
func CallUserSetSimInfo(c *Client, ctx context.Context, info *SimInfo) (errCode RpcErrCode) {
	var buf bytes.Buffer
	if err := SetAll(&buf, func(buf *bytes.Buffer) error {
		return SetSimInfo(buf, info)
	}); err != nil {
		return RpcReqErr
	}

	_, status := doClient(c, ctx, "/user.set_sim_info", buf.Bytes())
	if status != RpcOk {
		return status
	}

	return status
}
// CallGetCount 获取数量
func CallGetCount(c *Client, ctx context.Context, page uint8) (result uint8, errCode RpcErrCode) {
	var buf bytes.Buffer
	if err := SetAll(&buf, func(buf *bytes.Buffer) error {
		return SetU8(buf, page)
	}); err != nil {
		return result, RpcReqErr
	}

	body, status := doClient(c, ctx, "/get_count", buf.Bytes())
	if status != RpcOk {
		return result, status
	}

	if err := GetAll(bytes.NewBuffer(body), func(buf *bytes.Buffer) error {
		val, err := GetU8(buf)
		if err != nil { return err }
		result = val
		return nil
	}); err != nil {
		return result, RpcRespErr
	}
	return result, status
}
// CallGetBin 获取bin
func CallGetBin(c *Client, ctx context.Context, page uint8) (result []byte, errCode RpcErrCode) {
	var buf bytes.Buffer
	if err := SetAll(&buf, func(buf *bytes.Buffer) error {
		return SetU8(buf, page)
	}); err != nil {
		return result, RpcReqErr
	}

	body, status := doClient(c, ctx, "/get_bin", buf.Bytes())
	if status != RpcOk {
		return result, status
	}

	if err := GetAll(bytes.NewBuffer(body), func(buf *bytes.Buffer) error {
		val, err := GetBin(buf)
		if err != nil { return err }
		result = val
		return nil
	}); err != nil {
		return result, RpcRespErr
	}
	return result, status
}

File: ts/sb/rpc.ts
import * as _ from "./_"

const defaultMaxRespBytes = 4 * 1024 * 1024;
const maxSafeRespBytes = Number.MAX_SAFE_INTEGER;
const maxTimeoutMs = 2147483647;

export enum RpcErrCode {
    Ok = 200,
    NoConn = 0,
    Timeout = 408,
    ReqErr = 400,
    RespErr = 500,
    NotAuth = 401,
    NotExist = 404,
}

export interface RpcConfig {
    host: string;
    headers?: Record<string, string>;
    timeout?: number;
    retries?: number;
    maxRespBytes?: number;
}

export class RpcClient {
    private headers: Record<string, string> = {};
    private timeout: number;
    private retries: number;
    private maxRespBytes: number;

    constructor(private config: RpcConfig) {
        this.config = { ...config, host: config.host.replace(/\/+$/, "") };
        if (config.headers) this.headers = { ...config.headers };
        const cfgTimeout = config.timeout;
        this.timeout = cfgTimeout !== undefined && Number.isFinite(cfgTimeout) && cfgTimeout >= 0 ? Math.min(Math.floor(cfgTimeout), maxTimeoutMs) : 5000;
        const cfgRetries = config.retries;
        this.retries = cfgRetries !== undefined && Number.isFinite(cfgRetries) && cfgRetries >= 0 ? Math.floor(cfgRetries) : 3;
        const cfgMaxRespBytes = config.maxRespBytes;
        this.maxRespBytes = cfgMaxRespBytes !== undefined && Number.isFinite(cfgMaxRespBytes) && cfgMaxRespBytes > 0 ? Math.min(Math.floor(cfgMaxRespBytes), maxSafeRespBytes) : defaultMaxRespBytes;
    }

    public setHeader = (key: string, value: string): void => { this.headers[key] = value; };
    public getHeader = (key: string): string | undefined => this.headers[key];
    public removeHeader = (key: string): void => { delete this.headers[key]; };

    public setAuthorization = (token: string): void => { this.setHeader("Authorization", `Bearer ${token}`); };
    public getAuthorization = (): string | undefined => this.getHeader("Authorization");
    public removeAuthorization = (): void => { this.removeHeader("Authorization"); };
    public isAuthorized = (): boolean => !!this.getAuthorization();

    private async _fetch(path: string, body: Uint8Array): Promise<[Uint8Array | null, RpcErrCode]> {
        let lastStatus = RpcErrCode.NoConn;
        for (let i = 0; i <= this.retries; i++) {
            if (i > 0) await new Promise((res) => setTimeout(res, i * 1000));
            const controller = new AbortController();
            let timeoutId: ReturnType<typeof setTimeout> | null = null;
            if (this.timeout > 0) timeoutId = setTimeout(() => controller.abort(), this.timeout);
            try {
                const res = await fetch(`${this.config.host}/${path}`, {
                    method: "POST",
                    headers: { "Content-Type": "application/octet-stream", ...this.headers },
                    body: body as any,
                    signal: controller.signal,
                });
                if (res.ok) {
                    const contentLength = res.headers.get("content-length");
                    if (contentLength !== null) {
                        const size = Number(contentLength);
                        if (Number.isFinite(size) && size > this.maxRespBytes) return [null, RpcErrCode.RespErr];
                    }
                    const bytes = new Uint8Array(await res.arrayBuffer());
                    if (bytes.byteLength > this.maxRespBytes) return [null, RpcErrCode.RespErr];
                    return [bytes, RpcErrCode.Ok];
                }
                lastStatus = res.status as RpcErrCode;
                if (res.status === 408 && i < this.retries) continue;
                return [null, res.status as RpcErrCode];
            } catch (e: any) {
                lastStatus = e && e.name === "AbortError" ? RpcErrCode.Timeout : RpcErrCode.NoConn;
                if (i < this.retries) continue;
            } finally {
                if (timeoutId !== null) clearTimeout(timeoutId);
            }
        }
        return [null, lastStatus];
    }

    // 获取用户的id
    public userGetAbc = async (): Promise<[_.OrderStatus, RpcErrCode]> => {
        const buf = new _.Buffer();
        const [bytes, status] = await this._fetch("user.get_abc", buf.bytes);
        if (status !== RpcErrCode.Ok || bytes === null) return [_.DefaultOrderStatus(), status];
        const respBuf = new _.Buffer(bytes);
        let result = _.DefaultOrderStatus() as any;
        {
            const [value, err] = _.getU8(respBuf);
            if (err !== null) return [_.DefaultOrderStatus(), RpcErrCode.RespErr];
            const item = value as _.OrderStatus;
            if (!_.IsOrderStatus(item)) return [_.DefaultOrderStatus(), RpcErrCode.RespErr];
            result = item as any;
        }
        if (respBuf.len !== 0) return [_.DefaultOrderStatus(), RpcErrCode.RespErr];
        return [result as any, RpcErrCode.Ok];
    };

    // 获取abcd
    public userGetAbcd = async (page: number, size: number): Promise<[_.OrderStatus, RpcErrCode]> => {
        const buf = new _.Buffer();
        {
            const encodeErr = _.setU8(buf, page as any);
            if (encodeErr !== null) return [_.DefaultOrderStatus(), RpcErrCode.ReqErr];
        }
        {
            const encodeErr = _.setU8(buf, size as any);
            if (encodeErr !== null) return [_.DefaultOrderStatus(), RpcErrCode.ReqErr];
        }
        const [bytes, status] = await this._fetch("user.get_abcd", buf.bytes);
        if (status !== RpcErrCode.Ok || bytes === null) return [_.DefaultOrderStatus(), status];
        const respBuf = new _.Buffer(bytes);
        let result = _.DefaultOrderStatus() as any;
        {
            const [value, err] = _.getU8(respBuf);
            if (err !== null) return [_.DefaultOrderStatus(), RpcErrCode.RespErr];
            const item = value as _.OrderStatus;
            if (!_.IsOrderStatus(item)) return [_.DefaultOrderStatus(), RpcErrCode.RespErr];
            result = item as any;
        }
        if (respBuf.len !== 0) return [_.DefaultOrderStatus(), RpcErrCode.RespErr];
        return [result as any, RpcErrCode.Ok];
    };

    // 设置sim信息
    // 无返回值
    public userSetSimInfo = async (info: _.SimInfo): Promise<RpcErrCode> => {
        const buf = new _.Buffer();
        {
            const encodeErr = _.setSimInfo(buf, info);
            if (encodeErr !== undefined) return RpcErrCode.ReqErr;
        }
        const [bytes, status] = await this._fetch("user.set_sim_info", buf.bytes);
        if (status !== RpcErrCode.Ok || bytes === null) return status;
        if (bytes.byteLength !== 0) return RpcErrCode.RespErr;
        return RpcErrCode.Ok;
    };

    // 获取数量
    public getCount = async (page: number): Promise<[number, RpcErrCode]> => {
        const buf = new _.Buffer();
        {
            const encodeErr = _.setU8(buf, page as any);
            if (encodeErr !== null) return [0, RpcErrCode.ReqErr];
        }
        const [bytes, status] = await this._fetch("get_count", buf.bytes);
        if (status !== RpcErrCode.Ok || bytes === null) return [0, status];
        const respBuf = new _.Buffer(bytes);
        let result = 0 as any;
        {
            const [value, err] = _.getU8(respBuf);
            if (err !== null) return [0, RpcErrCode.RespErr];
            result = value as any;
        }
        if (respBuf.len !== 0) return [0, RpcErrCode.RespErr];
        return [result as any, RpcErrCode.Ok];
    };

    // 获取bin
    public getBin = async (page: number): Promise<[Uint8Array, RpcErrCode]> => {
        const buf = new _.Buffer();
        {
            const encodeErr = _.setU8(buf, page as any);
            if (encodeErr !== null) return [new Uint8Array(0), RpcErrCode.ReqErr];
        }
        const [bytes, status] = await this._fetch("get_bin", buf.bytes);
        if (status !== RpcErrCode.Ok || bytes === null) return [new Uint8Array(0), status];
        const respBuf = new _.Buffer(bytes);
        let result = new Uint8Array(0) as any;
        {
            const [state, errState] = _.getU8(respBuf);
            if (errState !== null) return [new Uint8Array(0), RpcErrCode.RespErr];
            const [value, err] = _.getBinCompact(respBuf, state);
            if (err !== null) return [new Uint8Array(0), RpcErrCode.RespErr];
            result = value as any;
        }
        if (respBuf.len !== 0) return [new Uint8Array(0), RpcErrCode.RespErr];
        return [result as any, RpcErrCode.Ok];
    };

}
